Framework/Algorithms/inc/MantidAlgorithms/SofQCommon.h

```diff
@@ -27,12 +27,20 @@ struct DLLExport SofQCommon {
   /// Get the efixed value for the given detector
   double getEFixed(const Geometry::IDetector &det) const;
 
+
+  /// Calculate the Q value
+  double q(const double deltaE, const double twoTheta, const Geometry::IDetector &det) const;
```

Comment:
Getting the detector from the instrument is a really expensive operation (one of the few that still needs to use the old instrument code), can't it be avoided in direct mode?